corehq/apps/hqadmin/management/commands/fix_checkpoints_from_date.py

```diff
@@ -0,0 +1,111 @@
+from __future__ import print_function
+
+import argparse
+import difflib
+import pprint
+import re
+from datetime import datetime
+
+from django.core.management import BaseCommand, CommandError
+
+from corehq.apps.hqadmin.models import HistoricalPillowCheckpoint
+from pillowtop.models import str_to_kafka_seq
+from pillowtop.utils import get_all_pillow_instances
+
+
+def valid_date(s):
+    try:
+        return datetime.strptime(s, "%Y-%m-%d")
+    except ValueError:
+        msg = "Not a valid date: '{0}'.".format(s)
+        raise argparse.ArgumentTypeError(msg)
+
+
+def confirm(msg):
+    input = raw_input("{} Type ['y', 'yes'] to continue.\n".format(msg))
+    if input in ['y', 'yes']:
+        return True
+
+
+class Command(BaseCommand):
+    help = ("Update the pillow sequence IDs based on a passed in file")
+
+    def add_arguments(self, parser):
+        parser.add_argument(
+            'date',
+            type=valid_date,
+            help="Take first checkpoint before date - format = YYYY-MM-DD",
+        )
+        parser.add_argument(
+            '--pillows',
+            nargs='+',
+            help="Restrict changes to only these pillows. (regex supported)"
+        )
+        parser.add_argument(
+            '--noinput',
+            action='store_true',
+            dest='noinput',
+            default=False,
+            help="Disable interactive mode",
+        )
+
+    def handle(self, **options):
+        date = options['date']
+        pillow_args = set(options['pillows'] or [])
+
+        if not pillow_args and not confirm('Reset checkpoints ALL pillows?'):
+            raise CommandError('Abort')
+
+        def _pillow_match(pillow_id):
+            return (
+                pillow_id in pillow_args
+                or any(re.match(arg, pillow_id, re.IGNORECASE) for arg in pillow_args)
+            )
+
+        all_pillows = get_all_pillow_instances()
+        if not pillow_args:
+            pillows = all_pillows
+        else:
+            pillows = [
+                pillow for pillow in all_pillows
+                if _pillow_match(pillow.pillow_id)
+            ]
+
+            if not pillows:
+                raise CommandError('No pillows match: {}'.format(options['pillows']))
+
+            if not confirm('Update checkpoints for {}?'.format(', '.join(p.pillow_id for p in pillows))):
```

Comment:
This might be easier to read later if it used '\n' to join and sorted the pillows by name (it may already do that in get all pillow instances though)

Reply:
4adcf6f